Stop formula engine

ela-kotulska-frequenz opened the pull request:
This PR introduces 4 new changes:
* make formula `stop` method public
* Add stop method to MetricFetcher - which stops receivers and FallbackMetricFetcher
* Fix bug that raised many exception when formulas stopped
* We can stop formula, now. So we can stop  FallbackMetricFetcher when primary metric fetcher is good again.

All this simplified the code a lot.

ela-kotulska-frequenz committed:
Make stop method of FormulaEngine public

To be able to stop custom formulas outside sdk.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Write debug instead of exception after FormulaEngine stop

Replace exception with debug message to avoid confusing traceback output.
Previously, FormulaEngine couldn't be stopped, so exception was acceptable.
Now that FormulaEngine can be stopped, use severity level 'debug' since
stopping is a normal operation and usefull only during debugging.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_engine.py:
```diff
@@ -313,7 +313,7 @@ async def _run(self) -> None:
                 msg = await evaluator.apply()
             except asyncio.CancelledError:
-                _logger.exception("FormulaEngine task cancelled: %s", self._name)
+                _logger.info("FormulaEngine task cancelled: %s", self._name)
```
Not sure if this is really useful information to have unless one is debugging. I would make it a `debug`.

ela-kotulska-frequenz replied:
Done

llucax reviewed (commented):
LGTM but the description about the `close` issue should be removed from the OR description so it doesn't go to the merge commit and it needs release notes as the formula engine is public.

ela-kotulska-frequenz commented:
One more error described in PR message

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:
```diff
@@ -529,7 +529,15 @@ async def _fetch_next(self) -> Sample[QuantityT] | None:
         try:
             next_value = await self._stream.receive()
         except ReceiverError[Any] as err:
```
My bet would be that `ReceiverError[Any]` is not really a `ReceiverError` but some `type` object or something like that. Maybe try replacing by just `ReceiverError`?

I just did a quick test and it is that:

```py
from typing import Any, Generic, TypeVar

T = TypeVar("T")

class Ex(Exception, Generic[T]):
    pass

try:
    raise Ex("foo")
except Ex[Any] as e:
    print(e)
```

This triggers the same issue, removing the `[Any]` makes it work.

ela-kotulska-frequenz replied:
You are right! :) It was ReceiverStoppedError .

llucax replied:
What I mean is the generic type specification was wrong, a `ReceiverStoppedError` is still a `ReceiverError`, but you can't use the type parameter in `except` clauses.

ela-kotulska-frequenz replied:
Right, sorry I didn't understand it. Fixed :)

ela-kotulska-frequenz committed:
Update release notes

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz commented:
Last commit is very complicated,  sorry for that!
But those things are really related and one implied the other and I couldn't separate it.

ela-kotulska-frequenz committed:
Add extra log about stop of fallback metric fetcher

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:
```diff
@@ -417,86 +433,69 @@ def stream_name(self) -> str:
+                    self._name,
+                )
+            return None
+        except ReceiverError[Any] as err:
```
This is still wrong, and if for any reason an exception that is not a `ReceiverStoppedError` is raised, we'll get the same error about the exception not inheriting from `BaseException`.

It should be:
```suggestion
        except ReceiverError as err:
```

ela-kotulska-frequenz replied:
Done

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:
```diff
@@ -506,34 +505,62 @@ async def fetch_next(self) -> Sample[QuantityT] | None:
+                "Failed to fetch next value from %s. Primary stream closed.",
+                self._name,
+            )
         except ReceiverError[Any] as err:
```
Even if this is a pr-existing bug, we should fix it too.
```suggestion
        except ReceiverError as err:
```

ela-kotulska-frequenz replied:
done

llucax reviewed (commented):
I'll leave the changes in formula engine to @shsms as he is more familiar with that code base.

ela-kotulska-frequenz committed:
Replace catching ReceiverError[Any] with just ReceiverError

It looks like ReceiverError[Any] is not ReceiverError, but some
type object. When any error occured we got missleading
error about the exception not inheriting from BaseException.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Refactor fetching from fallback_fetcher

Also simplify failure log message

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Rename `primary_fetcher_sample` → `primary_fetcher_value`

Also allow `None` values for it.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Refactor `_fetch_next` to prioritize the primary stream

When the primary stream has invalid data, then look at the fallback
stream.

This approach doesn't assume that once a fallback is started, it is
always running.

In a subsequent commit, we will implement stopping of fallback
fetchers when the primary has recovered.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Rename `next_value` → `primary_value`

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Remove unused `fetch_next_with_fallback` method

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Implement a `stop` method on `MetricFetcher`

And use it to stop the fallback fetcher when it is no longer needed.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Use a flag to identify if a `ReceiverStoppedError` is expected

If it happens because the fetcher is being stopped, then only log a
DEBUG message.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz commented:
Thanks @shsms for separating these commits. Is much more readable now! :)
